Produce profiles in google trace format (#17702)
coq/coq: doc/sphinx/practical-tools/coq-commands.rst

.. todo how about COQLIB, COQCORELIB, DOCDIR

> .. _COQ_PROFILE_COMPONENTS:

Author: This is how the other refs around this part are done

Control which components produce events when using the
:ref:`profiling` system. It should be a comma separated list of
components.

If the variable is not set, all components produce events.

> Components are internally defined, but `command` which corresponds to

Author: >Would be nice to include the current list of component names.

It would not be nice to have to maintain it.

:-list-tags: Print the highlight tags known by Coq as well as their
  currently associated color and exit.
:-h, --help: Print a short usage and exit.
:-time: Output timing information for each command to standard output.
:-time-file *file*: Output timing information for each command to the given file.
:-profile *file*: Output :ref:`profiling` information to the given file.

.. _profiling:

Profiling
---------

Using the `coqc` command line argument `-profile` or the environment
variable `PROFILE` with `coq_makefile`, profiling information is produced in
`Google trace format <https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit>`.

> The output consists of duration events for the execution of various
> components of Coq (for instance `process` for the whole file,
> `command` for each command, `pretyping` for elaboration).

Reviewer: ```suggestion
The output gives the duration and event counts for the execution of
components of Coq (for instance `process` for the whole file,
`command` for each command, `pretyping` for elaboration).
```

Maybe mention that some categories are included in others, eg command always includes parse_command time.

Author: >Maybe mention that some categories are included in others, eg command always includes parse_command time.

This could change in the future.

> Environment variable `COQ_PROFILE_COMPONENTS` can be used to filter

Reviewer: Hyperlink to the definition of the env variable

> the components which produce events. This may be needed to reduce the

Reviewer: ```suggestion
which components produce events. This may be needed to reduce the
```

> size of the produced file.

Reviewer: ```suggestion
size of the generated file.
```

> The produced file can be visualized for instance in

Reviewer: ```suggestion
The generated file can be visualized with
```

<https://ui.perfetto.dev> (which can directly load the `.gz`
compressed file produced by `coq_makefile`) or processed using any
JSON-capable system.

Events are annotated with additional information in the `args` field
(either on the beginning `B` or end `E` event):

> - `major` and `minor` indicate how many major and minor words were allocated during the event.

Reviewer: ```suggestion
- `major` and `minor` indicate how many major and minor words were allocated during all the events.
```

From the example, it looks like these allocation .counts apply to all events under `args`

Author: They are annotated on a specific event and correspond to that event. The presence of subevents is irrelevant.

> - `subtimes` indicates how much time was spent in sub-components and

Reviewer: How about including a short sample output file illustrating the various fields?

>   how many times each sub-component was profiled during the event

Reviewer: ```suggestion
  how many times each subcomponent was profiled during the event
```

  (including sub-components which do not appear in
  `COQ_PROFILE_COMPONENTS`).

> - for the `command` event, `cmd` displays the precise location of the

Reviewer: Would be great to show sample output for this.  Sounds like there is one `cmd` per unique command name.  Right?

Author: I don't know what you mean by "unique command name", `cmd` looks like `Chars 54 - 72 [Definition~a~:=~1.]`
